pkg/volume/csi/csi_plugin.go

```diff
@@ -263,72 +265,89 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
 		utilfeature.DefaultFeatureGate.Enabled(features.CSIMigration) {
 		// This function prevents Kubelet from posting Ready status until CSINodeInfo
 		// is both installed and initialized
-		initializeCSINodeInfo(host)
+		err := initializeCSINodeInfo(host)
+		if err != nil {
+			return fmt.Errorf("failed to initialize CSINodeInfo: %v", err)
+		}
 	}
 
 	return nil
 }
 
-func initializeCSINodeInfo(host volume.VolumeHost) {
+func initializeCSINodeInfo(host volume.VolumeHost) error {
+	kvh, ok := host.(volume.KubeletVolumeHost)
+	if !ok {
+		klog.V(4).Infof("Skipping CSINodeInfo initialization, not running on Kubelet")
+		return nil
+	}
 	kubeClient := host.GetKubeClient()
-	if kubeClient == nil || len(host.GetPodsDir()) == 0 {
+	if kubeClient == nil {
 		// If we're NOT on the Kubelet OR KubeClient DOESNT exist
-		klog.V(4).Infof("Skipping CSINodeInfo initialization, not running on Kubelet or KubeClient doesn't exist")
-		return
+		return fmt.Errorf("failed to get KubeClient")
 	}
 
+	kvh.SetKubeletError(fmt.Errorf("CSINodeInfo is not yet available."))
+
 	go func() {
-		host.SetKubeletError(fmt.Errorf("Could not find CSINodeInfo CRD. Please install the CSINodeInfo CRD"))
+		defer utilruntime.HandleCrash()
 
-		// TODO(dyzz) fine tune the backoff parameters for CRD Installation and Initialize migrated drivers
+		// Backoff parameters tuned to retry over 93 seconds. Will fail and restart the Kubelet
+		// after max retry steps.
 		crdBackoff := wait.Backoff{
-			Steps:    4,
+			Steps:    6,
 			Duration: 10 * time.Millisecond,
-			Factor:   5.0,
+			Factor:   6.0,
 			Jitter:   0.1,
 		}
-		for {
-			err := wait.ExponentialBackoff(crdBackoff, func() (bool, error) {
-				groupResource, err := kubeClient.Discovery().ServerResourcesForGroupVersion("csi.storage.k8s.io" + "/" + "v1alpha1")
-				if err != nil {
-					klog.Warningf("Failed to get groupResource: %v", err)
-					return false, nil
-				}
+		err := wait.ExponentialBackoff(crdBackoff, func() (bool, error) {
+			groupVersion := "csi.storage.k8s.io" + "/" + "v1alpha1"
+			groupResource, err := kubeClient.Discovery().ServerResourcesForGroupVersion(groupVersion)
+			if err != nil {
+				kvh.SetKubeletError(fmt.Errorf("Failed to get group resource for group version %v.", groupVersion))
+				klog.Errorf("Failed to get groupResource for group version %v: %v", groupVersion, err)
+				return false, nil
+			}
 
-				for _, g := range groupResource.APIResources {
-					if g.Name == "csinodeinfos" {
-						klog.V(4).Infof("Found the csinodeinfos CRD installed on API Server")
-						return true, nil
-					}
+			for _, g := range groupResource.APIResources {
+				if g.Name == "csinodeinfos" {
+					klog.V(4).Infof("Found the csinodeinfos CRD installed on API Server")
+					return true, nil
 				}
-				return false, nil
-			})
-			if err == nil {
-				break
 			}
+			return false, nil
+		})
+		if err != nil {
+			// Kill the Kubelet process and allow it to restart to retry initialization
+			os.Exit(1)
```

Comment:
I expected an error propagation, not an os.Exit(1) buried deep in init... do we have other examples of init loops in the kubelet triggering exits? if we can't propgate an error, at least exit via `klog.Fatalf` so we get a message indicate what caused the exit

Reply:
since this is inside a `go func` I don't think error propagating is an option here. `klog.Fatalf` seems like the right choice